hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java

```diff
@@ -0,0 +1,190 @@
+ */
+package org.hibernate.engine.jdbc.connections.internal;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.jboss.logging.Logger;
+
+import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider;
+import org.hibernate.service.Service;
+import org.hibernate.service.ServiceRegistry;
+import org.hibernate.service.spi.Configurable;
+import org.hibernate.service.spi.ServiceRegistryAwareService;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+import org.hibernate.service.spi.Stoppable;
+
+/**
+ * {@link ConnectionProvider} delegating work to a real connection provider
+ * and executing some operations when the first connection request returns a valid
+ * connection.
+ *
+ * Useful to make sure an application can start even if the database is nto yet ready or present
+ * while still be able to execute schema generation orders.
+ *
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public class DelayerConnectionProvider implements
+		ConnectionProvider,
+		Configurable,
+		ServiceRegistryAwareService,
+		Stoppable {
+
+	private static final Logger LOGGER = Logger.getLogger( DelayerConnectionProvider.class );
+
+	private final ConnectionProvider connectionProvider;
+	private ServiceRegistry serviceRegistry;
+	// volatile due to the double checked lock
+	private volatile List<Runnable> operations = new ArrayList<>();
+	private final ReentrantLock lock = new ReentrantLock();
+
+	public DelayerConnectionProvider(ConnectionProvider connectionProvider) {
+		this.connectionProvider = connectionProvider;
+	}
+
+	/**
+	 * Operations can only be added before the {@code SessionFactory} is fully built.
+	 * This is to prevent concurrent additions of operations.
+	 * The operation must use the {@link ServiceRegistry} provided to retrieve
+	 * {@link ConnectionProvider} to get a connection.
+	 *
+	 * The operation will try and be executed eagerly, we might be lucky.
+	 *
+	 * @param operation on a connection ready database
+	 */
+	public void addOperation(Runnable operation) {
+		operations.add( operation );
+		tryAndExecuteOperations();
+	}
+
+	/**
+	 * Try to eagerly flush the operations if a connection is available
+	 */
+	private void tryAndExecuteOperations() {
+		// It would be more elegant to refactor the operation lock from getConnection is a shared method
+		// but I prefer the JVM to have more chances to keep getConnection() inlined
+		try {
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
+			}
+		}
+		catch (Exception e) {
+			// we don't care if there is a failure
```

Comment:
I would log the exception, with a Map<String,Runnable> instead of a List<Runnable> where the key is the name of the operation we could log also which operation failed